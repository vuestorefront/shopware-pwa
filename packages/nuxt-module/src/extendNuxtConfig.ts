import { NuxtConfig } from "@nuxt/types";
import { isArray, mergeWith } from "lodash";

function customizer(objValue: Object, srcValue: unknown) {
  if (isArray(objValue)) {
    return objValue.concat(srcValue);
  }
}

const defaultConfig: NuxtConfig = {
  telemetry: false,
  vue: {
    config: {
      devtools:
        process.env.ENABLE_DEVTOOLS === "true" ||
        process.env.NODE_ENV !== "production",
    },
  },
  env: {
    CHOKIDAR_USEPOLLING: process.env.NODE_ENV == "production" ? "0" : "1",
    EXPERIMENTAL_IMAGE_PROCESSING_SERVER:
      process.env.EXPERIMENTAL_IMAGE_PROCESSING_SERVER || "",
    NODE_ENV: process.env.NODE_ENV || "",
    ENABLE_DEVTOOLS: process.env.ENABLE_DEVTOOLS || "",
  },
  server: {
    port: process.env.PORT || 3000,
    host: process.env.HOST || "0.0.0.0",
  },
  // Global page headers: https://go.nuxtjs.dev/config-head
  head: {
    title: "Shopware PWA",
    meta: [
      { hid: "project-type", name: "project-type", content: "shopware-pwa" },
      { charset: "utf-8" },
      { name: "viewport", content: "width=device-width, initial-scale=1" },
      { hid: "description", name: "description", content: "" },
    ],
    link: [
      { rel: "icon", type: "image/x-icon", href: "/favicon.ico" },
      // TODO: remove this, only for default-theme
      {
        rel: "preload",
        href: "https://fonts.googleapis.com/css?family=Raleway:300,400,400i,500,600,700%7CRoboto:300,300i,400,400i,500,700&display=swap",
        as: "style",
      },
      {
        rel: "stylesheet",
        href: "https://fonts.googleapis.com/css?family=Raleway:300,400,400i,500,600,700%7CRoboto:300,300i,400,400i,500,700&display=swap",
        media: "print",
        onload: "this.media='all'",
        once: true,
      },
    ],
  },

  // Global CSS: https://go.nuxtjs.dev/config-css
  // TODO: remove this, only for default-theme
  css: ["~assets/scss/main.scss"],
  // css: [],

  // Plugins to run before rendering page: https://go.nuxtjs.dev/config-plugins
  plugins: [],

  // Auto import components: https://go.nuxtjs.dev/config-components
  components: false,

  // Modules for dev and build (recommended): https://go.nuxtjs.dev/config-modules
  buildModules: ["@nuxt/typescript-build", "@shopware-pwa/nuxt-module"],

  // Modules: https://go.nuxtjs.dev/config-modules
  modules: [
    // https://go.nuxtjs.dev/axios
    "@nuxtjs/axios",
    // https://go.nuxtjs.dev/pwa
    "@nuxtjs/pwa",
  ],

  // Axios module configuration: https://go.nuxtjs.dev/config-axios
  axios: {},

  // PWA module configuration: https://go.nuxtjs.dev/pwa
  pwa: {
    manifest: {
      lang: "en",
    },
    workbox: {
      offlineStrategy: "StaleWhileRevalidate",
    },
  },

  // Build Configuration: https://go.nuxtjs.dev/config-build
  build: {
    babel: {
      plugins: [["@babel/plugin-proposal-private-methods", { loose: true }]],
    },
  },
  render: {
    bundleRenderer: {
      runInNewContext: false,
    },
  },
};

const configs: NuxtConfig[] = [defaultConfig];
export default function extendNuxtConfig(config: NuxtConfig) {
  configs.push(config);
  return configs.reduce((prev, next) => mergeWith(prev, next, customizer));
}
