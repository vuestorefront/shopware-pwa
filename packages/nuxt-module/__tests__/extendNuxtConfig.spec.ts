import extendNuxtConfig from "../src/extendNuxtConfig";

describe("nuxt-module - extendNuxtConfig", () => {
  it("should return default config keys", () => {
    const result = extendNuxtConfig({});
    expect(Object.keys(result)).toMatchSnapshot();
  });

  it("should overwrite server port", () => {
    const result = extendNuxtConfig({
      server: {
        port: 2000,
      },
    });
    expect(result.server?.port).toEqual(2000);
  });

  it("should add new build module", () => {
    const result = extendNuxtConfig({
      buildModules: ["some-new-module"],
    });
    expect(result.buildModules).toEqual([
      "@nuxt/typescript-build",
      "@shopware-pwa/nuxt-module",
      "some-new-module",
    ]);
  });

  describe("env properties", () => {
    const ORIGINAL_ENV = process.env;

    beforeEach(() => {
      jest.resetModules(); // clears the cache
      process.env = { ...ORIGINAL_ENV };
    });

    afterAll(() => {
      process.env = ORIGINAL_ENV;
    });

    it("should have default env values", () => {
      process.env.NODE_ENV = undefined;

      const extendNuxtConfig = require("../src/extendNuxtConfig").default;
      const result = extendNuxtConfig({});

      expect(result.env?.CHOKIDAR_USEPOLLING).toEqual("1");
      expect(result.env?.NODE_ENV).toEqual("");
      expect(result.env?.ENABLE_DEVTOOLS).toEqual("");
    });

    it("should pass env properties to config", () => {
      process.env.NODE_ENV = "production";
      process.env.EXPERIMENTAL_IMAGE_PROCESSING_SERVER = "https://example.com";
      process.env.ENABLE_DEVTOOLS = "true";

      const extendNuxtConfig = require("../src/extendNuxtConfig").default;
      const result = extendNuxtConfig({});

      expect(result.env?.CHOKIDAR_USEPOLLING).toEqual("0");
      expect(result.env?.EXPERIMENTAL_IMAGE_PROCESSING_SERVER).toEqual(
        "https://example.com"
      );
      expect(result.env?.NODE_ENV).toEqual("production");
      expect(result.env?.ENABLE_DEVTOOLS).toEqual("true");
    });
  });
});
