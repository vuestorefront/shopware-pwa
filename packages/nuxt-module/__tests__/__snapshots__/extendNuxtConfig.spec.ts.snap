// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nuxt-module - extendNuxtConfig should return default config keys 1`] = `
Array [
  "telemetry",
  "vue",
  "env",
  "server",
  "head",
  "css",
  "plugins",
  "components",
  "buildModules",
  "modules",
  "axios",
  "pwa",
  "build",
  "render",
]
`;
